python/cudf_polars/cudf_polars/experimental/rapidsmpf/join.py

```diff
@@ -180,28 +176,72 @@ async def broadcast_join_node(
                     ),
                 )
             allgather.insert_finished()
+            stream = ir_context.get_cuda_stream()
             small_dfs = [
                 DataFrame.from_table(
-                    await allgather.extract_concatenated(small_df.stream),
+                    await allgather.extract_concatenated(stream),
                     list(small_child.schema.keys()),
                     list(small_child.schema.values()),
-                    small_df.stream,
+                    stream,
                 )
             ]
+        elif small_dfs and (
+            ir.options[0] != "Inner" or small_size < target_partition_size
+        ):
+            # Pre-concat for non-inner joins, otherwise
+            # we need a local shuffle, and face additional
+            # memory pressure anyway.
+            small_dfs = [_concat(*small_dfs, context=ir_context)]
 
         # Stream through large side, joining with the small-side
-        while (msg := await large_ch.data.recv(context)) is not None:
-            large_chunk = TableChunk.from_message(msg).make_available_and_spill(
-                context.br(), allow_overbooking=True
-            )
-            seq_num = msg.sequence_number
+        seq_num = 0
+        large_chunk_processed = False
+        receiving_large_chunks = True
+        while receiving_large_chunks:
+            msg = await large_ch.data.recv(context)
+            if msg is None:
+                receiving_large_chunks = False
+                if large_chunk_processed:
+                    # Normal exit - We've processed all large-table data
+                    break
+                elif small_dfs:
+                    # We received small-table data, but no large-table data.
+                    # This may never happen, but we can handle it by generating
+                    # an empty large-table chunk
+                    stream = ir_context.get_cuda_stream()
+                    large_chunk = empty_table_chunk(large_child, context, stream)
```

Comment:
Are we missing a `break` in this elif? IIUC, we'll go back to `msg = await large_ch.data.recv(context)`, which should hang.

Reply:
No, we will end up on line 225 after this, where we will use empty data in our join. Then the next tile we check the while condition, receiving_large_chunks will be False.